cmd/horcrux/cmd/state.go

```diff
@@ -103,6 +117,84 @@ func setStateCmd() *cobra.Command {
 	}
 }
 
+func importStateCmd() *cobra.Command {
+	return &cobra.Command{
+		Use:     "import [height]",
+		Aliases: []string{"i"},
+		Short: "Read the old priv_validator_state.json and set the height, round and step" +
+			"(good for migrations but NOT shared state update)",
+		Args:         cobra.ExactArgs(0),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if _, err := os.Stat(config.HomeDir); os.IsNotExist(err) {
+				cmd.SilenceUsage = false
+				return fmt.Errorf("%s does not exist, initialize config with horcrux config init and try again", config.HomeDir)
+			}
+
+			// Resetting the priv_validator_state.json should only be allowed if the
+			// signer is not running.
+			if err := signer.RequireNotRunning(config.PidFile); err != nil {
+				return err
+			}
+
+			pv, err := signer.LoadSignState(config.privValStateFile(config.Config.ChainID))
+			if err != nil {
+				return err
+			}
+
+			// Allow user to paste in priv_validator_state.json
+
+			fmt.Println("IMPORTANT: Your validator should already be STOPPED.  You must copy the latest state..")
+			<-time.After(2 * time.Second)
+			fmt.Println("")
+			fmt.Println("Paste your old priv_validator_state.json.  Input a blank line after the pasted JSON to continue.")
+			fmt.Println("")
+
+			var textBuffer strings.Builder
+
+			scanner := bufio.NewScanner(os.Stdin)
+			for scanner.Scan() {
+				if len(scanner.Text()) == 0 {
+					break
+				}
+				textBuffer.WriteString(scanner.Text())
+			}
+			finalJSON := textBuffer.String()
+
+			pvState := &FilePVLastSignState{}
+
+			err = tmjson.Unmarshal([]byte(finalJSON), &pvState)
+			if err != nil {
+				fmt.Println("Error parsing priv_validator_state.json")
+				return err
+			}
+
+			pv.EphemeralPublic = nil
+			signState := signer.SignStateConsensus{
+				Height:    pvState.Height,
+				Round:     int64(pvState.Round),
+				Step:      pvState.Step,
+				Signature: nil,
+				SignBytes: nil,
+			}
+			fmt.Printf("Saving New Sign State: \n"+
+				"  Height:    %v\n"+
+				"  Round:     %v\n"+
+				"  Step:      %v\n",
+				signState.Height, signState.Round, signState.Step)
+
+			err = pv.Save(signState, nil)
+			if err != nil {
+				fmt.Printf("error saving Sign State")
+				return err
+			}
+			<-time.After(100 * time.Millisecond) // some machines are too fast
```

Comment:
This points out an issue with the async nature of the `(*SignState) Save` method. I think rather than waiting an amount of time, we should modify the method to take a third parameter such as `async bool`, and it can do the `signState.save()` call in the goroutine if `async` is true. Then we could pass true everywhere except for `setStateCmd` and `importStateCmd`

Reply:
Since you did not introduce this bug, I created a PR into your branch here https://github.com/chillyvee/horcrux/pull/1

Reply:
A PR into my PR? Wow!

Reply:
Merged but checking operation manually.

Reply:
Seems to be an issue on first import when shared_state file does not exist yet.

Reply:
Adding commit to create file if it doesn't exist